Opts plugin

kodebach opened the pull request:
Closes #2427

## Basics

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] meta data is updated (e.g. README.md of plugins)

## Review

@markus2330 please review my pull request

kodebach commented on src/plugins/gopts/gopts_win32.h:
```diff
@@ -0,0 +1,100 @@
+{
+	int argc;
+
+	LPWSTR * args = CommandLineToArgvW (GetCommandLineW (), &argc);
```
This implementation is a bit weird and I'm not entirely sure it will work if any non ASCII characters are passed as command-line arguments, but it is the best I could do.

The only alternative is using `GetCommandLine` (will use ASCII mode when `UNICODE` is not defined), but because there is no non-Unicode `CommandLineToArgv`, we would have to parse the string ourselves. Because ASCII mode in Windows isn't necessarily ASCII, but rather any 8-bit character encoding, that could be even more difficult than it already is (quotes, escaping, etc.)

markus2330 replied:
Should be okay for now.

markus2330 commented on src/include/kdbconfig.h.in:
```diff
@@ -200,8 +200,25 @@
 #cmakedefine HAVE_HSEARCHR
 #endif
 
+/* define if the gopts plugin shall use the procfs implementation */
```
Better define this only locally for your plugin.

markus2330 commented on src/plugins/gopts/config.c.in:
```diff
@@ -0,0 +1,10 @@
+ *
+ */
+
+#define TESTAPP_PATH "@TESTAPP_PATH@"
```
(here)

markus2330 commented on src/plugins/gopts/gopts.c:
```diff
@@ -0,0 +1,82 @@
+#elif defined(ELEKTRA_GOPTS_SYSCTL)
+#include "gopts_sysctl.h"
+#else
+// define anyway to get rid of IDE warnings
```
define it before (as prototype)?

kodebach commented:
The more I work on this, the more I realize, this plugin won't work the way we intended it to. At least not without some major modifications to the core of Elektra.

1. The plugin really only makes sense, if it is executed between `getstorage` and `postgetstorage`. Otherwise we bypass some or all of the validation that is done in `postgetstorage`.
2. The plugin doesn't actually make sense as a global plugin. It has no need to see the whole KDB. In fact, in an ideal world, it would only see the specification of the application currently calling `kdbGet`. This isn't possible (maybe it is? see idea below) so the next best thing is, mounting the plugin automatically for every mountpoint. But because the number of plugins foreach mountpoint is very limited right now, this isn't a good solution either.
3. We already knew that some applications (like `kdb`) will need to disable the plugin, because they access parts of the KDB that might contain other specifications. This again requires modifications of the core of Elektra.

All of these modifications to the core might be possible, but considering that there are some major changes to the plugin framework planned and most of the changes required for this plugin might interfere with these, I don't think now is the right time to implement this plugin.

---

What might actually solve all these problems, is the `specload` plugin. Because it is mounted as `getstorage` all the validation will be done as expected. The `specload` plugin knows which application the specification belongs to (given by the `app` config key). It could compare the current executables path (can be obtained by methods similar to the ones `gopts` uses) and if they match it invokes `gopts`.

The only problem is, I don't know how the plugin framework handles, if a plugin returns Keys outside the given parentKey.

markus2330 commented:
> The plugin really only makes sense, if it is executed between getstorage and postgetstorage. Otherwise we bypass some or all of the validation that is done in postgetstorage.

We can add further places. Or we could use infos/ordering to make the validation earlier.

>  It has no need to see the whole KDB. In fact, in an ideal world, it would only see the specification of the application currently calling kdbGet

Actually, you can mount plugins to spec (only). Spec is excluded from cascading mountpoints on purpose.

> We already knew that some applications (like kdb) will need to disable the plugin, because they access parts of the KDB that might contain other specifications. This again requires modifications of the core of Elektra.

I think such applications can also simply throw the content from proc/.

> The only problem is, I don't know how the plugin framework handles, if a plugin returns Keys outside the given parentKey.

I think the framework currently throws the keys away. But I also think that it somehow makes sense to have proc/ as exception as it is no ordinary namespaces (no plugin is responsible for it).

kodebach commented:
@markus2330 The plugin is now added as a global plugin by default (like `spec`). However, I disabled it by default for now. I added a very crude function to enable it for testing.

My next step will be adding `kdbEnsure`. However, to do that I need to know, what functionality it should support and how generic the contract should be? Should it only support mounting/unmountig global plugins or should it also replace the `notification` library? (Which AFAIK is only there to setup the `internalnotification` plugin)

If `kdbEnsure` should be able to do more than mounting/unmounting, I think we need some interface to the plugins, so the plugin specific code stays isolated...

markus2330 commented:
This pull request **introduces 2 alerts** when merging 7370a5f7cd75f0229f641307cfc012c6b0f9844d into 05a3613957f87d4f3bcf90add539b9af65981426 - [view on LGTM.com](https://lgtm.com/projects/g/ElektraInitiative/libelektra/rev/pr-6702fe6e58d3a9395b26d1eb784106e5379fa030)

**new alerts:**

* 2 for Missing return statement

---

*Comment posted by [LGTM.com](https://lgtm.com)*

markus2330 commented:
This pull request **introduces 2 alerts** when merging 844f566e958df7e21af6331530c54369a6614c4d into 05a3613957f87d4f3bcf90add539b9af65981426 - [view on LGTM.com](https://lgtm.com/projects/g/ElektraInitiative/libelektra/rev/pr-1575ca84870a493f13112d58383701bac5e8a78a)

**new alerts:**

* 2 for Missing return statement

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kodebach commented:
jenkins build all please

kodebach commented:
I think this is done now...

@markus2330 please review

mpranj commented:
Why is the jenkins build green/OK when there is [one terminated/failed build](https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/detail/PR-2471/22/pipeline/449)?

mpranj commented on doc/news/_preparation_next_release.md:
```diff
@@ -51,7 +51,25 @@ the `type` plugin will ignore the key. We now also support converting enum value
+contract. In principle this a very powerful tool that may be used for a lot of things. For now it only supports a few conditions concerning
+plugins:
+
+- You can specify that a plugin should be mounted globally. This is can for example be used to enable the new [gopts](#gopts) plugin.
```
"This **is can** for example" -> "This can for example"

mpranj commented on doc/news/_preparation_next_release.md:
```diff
@@ -171,7 +198,7 @@ compiled against an older 0.8 version of Elektra will continue to work
 ### Core
 
-- <<TODO>>
+- `kdbGet` now calls global postgetstorage plugins with the parent key passed to `kdbGet`, instead of a random mountpoint. _(Klemens Böswirth)_
```
👍

kodebach replied:
One thing I noticed (but didn't look into): At least some errors returned from global plugins are suppressed by `kdbGet`. You may want to look into that while working on #2457.

mpranj replied:
Yep, we didn't even use return values until #2307. The global plugin implementation needs to be re-worked and tested properly, so that it matches the proposal.

kodebach commented:
> Why is the jenkins build green/OK when there is one terminated/failed build?

I guess only the Jenkins devs could answer that one ....

kodebach commented:
jenkins build libelektra please

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -51,7 +51,25 @@ the `type` plugin will ignore the key. We now also support converting enum value
-### <<HIGHLIGHT2>>
+### kdbEnsure
+
+`kdbEnsure` is a new function in `elektra-kdb`. It can be used to ensure that a KDB instance meets certain conditions specified in a
```
conditions -> clauses (2x)

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -51,7 +51,25 @@ the `type` plugin will ignore the key. We now also support converting enum value
+- Conversely you can also define that a plugin should not be mounted globally, e.g. to disable the `spec` plugin, which is enabled by default.
+- Additionally you may want to enforce that a global plugin uses a certain configuration. For this case you can specify that the plugin
+  should be remounted, i.e. unmounted and immediately mounted again.
+- In future non-global plugins will support the same features. But because of the different architecture involved, for now only unmounting
```
No need to mention it. We might or might not add it.

markus2330 commented on src/bindings/cpp/include/kdbexcept.hpp:
```diff
@@ -60,10 +60,30 @@ class KDBException : public Exception
-private:
+protected:
 	Key m_key;
 	mutable std::string m_str;
```
Could be private?

markus2330 commented on src/libs/elektra/kdb.c:
```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+		ksDel (pluginConfig);
+		return ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey);
+	case PLUGIN_STATE_MOUNTED:
+		return 0; // TODO: ensurePluginMounted (handle, mountpoint, pluginName, pluginConfig, errorKey);
```
Also assert here that this is not called? (2x)

markus2330 commented on src/libs/elektra/kdb.c:
```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+ * If `<mountpoint>` is NOT `global`, currently only `unmounted` is supported (not `mounted` and `remounted`).
+ *
+ * NOTE: This function only works properly, if the list plugin is mounted in all global positions.
+ * If this is not the case, 1 will be returned, because this is seen as an implicit condition in the contract.
```
Better to fail if you cannot do it. The application has no control over if the list plugin is there or not. And it expects that `kdbEnsure` actually ensured its contract if it returned successfully.

kodebach replied:
The return value 1 indicates a failure. Only 0 means success.

markus2330 replied:
This is non-standard compared to other kdb calls. Can you change that to be -1?

kodebach replied:
I could change it to -2, if you want all negative return values + 0 for success...

1 is returned, if the contract could not ensured, but the `kdbEnsure` call was entirely valid. e.g. if a plugin could not be mounted

-1 meanwhile is used to indicate that either the contract is malformed, or something else about the `kdbEnsure` call is wrong.

If 1 is returned, the application should show a meaningful message to the user, because they might need to change their KDB configuration. If -1 is returned, the application probably has a bug, and without changing the code there is little to no chance of fixing the problem...

markus2330 replied:
> I could change it to -2, if you want all negative return values + 0 for success...

I do not think we need this distinction in the return value as you can simply set different errors in the errorKey.

Furthermore, I think it might be useful to indicate if something was done at all.
So I would propose following return codes:

- 1: could not run kdbEnsure successfully (error is in the errorKey)
0: kdbEnsure successful but nothing was to do
1: mounted or unmounted something, so the application now has different mountpoints than the rest of the system

This would be basically identical to return codes of the other kdb* calls (and also the resolver).

Iirc there are now also macros defined for these values.

markus2330 commented on src/libs/elektra/kdb.c:
```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+ * This function can be used the given KDB @p handle meets certain conditions,
+ * specified in @p contract. Currently the following conditions are supported:
+ *
+ * - `system/plugins/<mountpoint>/<pluginname>` defines the state of the plugin
```
What about the name `system/elektra/ensure/plugins/...`? `system/plugins` is not reserved and it will create problems once these keys should be read from KDB.

kodebach replied:
I can change it, but I don't see a reason, why these keys should be stored in the KDB. They are not configuration, they are static specification. IMO the use case for `kdbEnsure` is that an application wants to make sure the KDB is in a state that is usable to the application. Changing this contract should only be necessary, if the code changes, i.e. recompiling is necessary.

That is also the reason, why I decided it is okay to consume the contract. If the keys came from the KDB this would be rather inconvenient.

markus2330 replied:
> I can change it, but I don't see a reason, why these keys should be stored in the KDB.

It is not about storing them there but it might be useful to be able to query the contracts of applications (with a functionality similar to the specload).

markus2330 commented on src/libs/elektra/kdb.c:
```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+ *
+ * @param handle    contains internal information of @link kdbOpen() opened @endlink key database
+ * @param contract  KeySet containing the contract described above.
+ *                  Because of the implementation, this will always be `ksDel()`ed. <b>Even in error cases.</b>
```
" Because of the implementation" is redundant information. Maybe you wanted to say: "To allow `kdbEnsure(.., ksNew(...)`..."

kodebach replied:
> Maybe you wanted to say: "To allow kdbEnsure(.., ksNew(...)..."

That is a nice side effect, but in most cases won't make for good, readable code. The actual reason is that I use `ksCut` and wanted to avoid having to `ksDup` first, because the contract shouldn't be needed after calling `kdbEnsure` anyway.

markus2330 commented on src/tools/kdb/CMakeLists.txt:
```diff
@@ -20,7 +20,7 @@ if (BUILD_SHARED)
 	add_dependencies (kdb kdberrors_generated elektra_error_codes_generated)
 
-	target_link_libraries (kdb elektra-core elektra-kdb elektratools)
+	target_link_libraries (kdb elektra-core elektra-kdb elektratools elektra-proposal)
```
This is ugly and should be avoided. Which methods from proposal do you need?

kodebach replied:
I think, this is actually just an artifact from some earlier tests, because I didn't actually change the `kdb` tool. I just used it for testing.

markus2330 replied:
Ok, perfect, then you can simply change it back.

markus2330 commented on src/plugins/gopts/README.md:
```diff
@@ -0,0 +1,34 @@
+
+## Usage
+
+The preferred way of using this plugin is via `kdbEnsure`:
```
Isn't global mounting possible, too?

kodebach replied:
Technically yes, but the `kdb` tool doesn't call `kdbEnsure` yet (to make sure `gopts` is unmounted, when accessing third-party configuration), so `kdb get` etc may break in certain cases.

`gopts` is unmounted by default for now. My plan was to make it mounted by default in future and just unmount it in `kdb`. That is however quite tedious work, because each command uses its own `KDB` instance, not created through some shared means.

markus2330 replied:
What about subclassing KDB (where `gopts` is ensured to be not there) and replace all KDB with the subclass? (Should be a simply search&replace.)

kodebach replied:
I think it would be better to create `KDBCommand`, an abstract subclass of `Command` that provides a `KDB` instance to its subclasses. I think this is a nicer solution, because later on KDB should use `gopts` for its own config and only the commands accessing the rest of the KDB should use the version without `gopts`.

I will modify `kdb` in a follow-up PR and then we might mount `gopts` by default.

markus2330 commented on src/plugins/gopts/README.md:
````diff
@@ -0,0 +1,34 @@
+The preferred way of using this plugin is via `kdbEnsure`:
+
+```c
+KDB * kdb = kdbOpen (parentKey);
````
It would be nice if this code is also as stand-alone minimal-working example available.

markus2330 commented:
Btw. did you also try if `kdbEnsure` is available in the non-C++ bindings? (E.g. Python) It would be great if cmd-line parsing would also be available from within Python.

sanssecours commented on src/plugins/list/README.md:
````diff
@@ -48,6 +48,23 @@ A list of set-placements for the plugin. Same for "get" and "error"
+Plugin * elektraListFindPlugin (Plugin * handle, const char * pluginName)
+```
+
+`elektraListMountPlugin` can be used to add a new plugin to the config. The placement will be query from the plugin it self (from its
````
Did you maybe mean “queried” instead of “query”?

sanssecours replied:
As far as I know, “it self” should be written as [one word](https://www.merriam-webster.com/dictionary/itself) in this context.

sanssecours commented on src/plugins/gopts/testapp.c:
```diff
@@ -0,0 +1,112 @@
+{
+	if (strcmp (name, TEST_EMPTY) == 0)
+	{
+		*parentKey = keyNew ("spec/tests/gopts", KEY_END);
```
Maybe it would make sense to create the parent key only once. As far as I can tell, each of the `if`-statements creates exactly the same key.

sanssecours commented on src/plugins/list/README.md:
```diff
@@ -48,6 +48,23 @@ A list of set-placements for the plugin. Same for "get" and "error"
 
 Plugin specific config.
 
+## Exported functions
```
Please use [title case](https://www.grammar-monster.com/lessons/capital_letters_title_case.htm) for all headings.

markus2330 replied:
I think this needs to be added to doc/CODING.md?

kodebach commented on src/bindings/cpp/include/kdb.hpp:
```diff
@@ -232,6 +234,33 @@ inline int KDB::set (KeySet & returned, Key & parentKey)
+int KDB::ensure (const KeySet & contract, Key & parentKey)
+{
+	// have to ksDup because contract is consumed and ksDel()ed by kdbEnsure
+	int ret = ckdb::kdbEnsure (handle, ckdb::ksDup (contract.getKeySet ()), parentKey.getKey ());
```
No that won't work, because that will cause a double free, because the C++ class KeySet will call `ksDel()` in the destructor.

kodebach commented:
> Btw. did you also try if `kdbEnsure` is available in the non-C++ bindings?

It is available in Python (tested) and should be available in any binding that is auto-generated.

> It would be great if cmd-line parsing would also be available from within Python.

It is technically possible right now, but because `argv[0]` is the python executable the remaining args will also contain the python script file, and the help message will contain the wrong program name. I have a solution for that, but I will put it into a follow-up PR. Together with a full Python example.

markus2330 commented on examples/gopts.c:
```diff
@@ -0,0 +1,241 @@
+		KS_END);
+}
+
+static int setupSpec (void)
```
Please mention here that this is only like this to be self-contained, not that an application should do this.

markus2330 commented on src/plugins/multifile/README.md:
```diff
@@ -102,7 +102,7 @@ kdb ls user/tests/multifile
 kdb rm -r user/tests/multifile/test.ini
 
-stat ~/.config/multifile/test.ini
+stat $(dirname $(kdb file user))/multifile/test.ini
```
We also should add build jobs which use non-standard config paths (ideally also with some umlauts, white spaces, ...)

kodebach replied:
AFAIK most if not all build jobs (certainly the Jenkins ones) use non-standard config paths inside the workspace directory. I also run all local tests with a similar setup.

About umlauts/white-space I am not sure, we might want to create an issue

markus2330 replied:
So the multifile test cases are not running on our build server?

kodebach replied:
I just checked and at least the `debian-stable-full` build runs this test successfully... Maybe some of the changes to `kdbGet` I made, brought up a hidden bug which was suppressed before

markus2330 replied:
Thank you for looking into it!

kodebach commented:
@markus2330 I think we can merge this, if not, please tell me what I missed

markus2330 commented:
Thank you so much for this very nice PR! It is great that not only LCDproc will from now have command-line arguments.

@ulschaef hopefully will also soon have finished the command-line argument completion, which would make our solution complete :+1:
